packages/xod-client/src/editor/containers/Patch/modes/marqueeSelecting.jsx

```diff
@@ -0,0 +1,218 @@
+  const x = Math.min(startPos.x, endPos.x);
+  const y = Math.min(startPos.y, endPos.y);
+  const x1 = Math.max(startPos.x, endPos.x);
+  const y1 = Math.max(startPos.y, endPos.y);
+  const width = Math.abs(startPos.x - endPos.x);
+  const height = Math.abs(startPos.y - endPos.y);
+
+  return ({
+    from: { x, y },
+    to: { x: x1, y: y1 },
+    width,
+    height,
+  });
+};
+
+const findSelectedItems = (api) => {
+  const inclusive = isInclusiveSelection(api);
+  const selectionBox = getSelectionBox(api);
+
+  const filterLinksFn = (
+    inclusive ?
+    filterLinksByInclusiveBox :
+    filterLinksByBox
+  )(selectionBox);
+  const filterNodesFn = (
+    inclusive ?
+    filterNodesByInclusiveBox :
+    filterNodesByBox
+  )(selectionBox);
+
+  return R.compose(
+    R.evolve({
+      links: filterLinksFn,
+      nodes: filterNodesFn,
+      comments: filterNodesFn,
+    }),
+    R.map(R.values),
+    R.pick(['links', 'nodes', 'comments'])
+  )(api.props);
+};
+
+const getComputedSelection = (api, event) => {
+  const oldSelection = api.props.selection;
+  const newSelection = getNewSelection(findSelectedItems(api));
+
+  return (
+    !isSelectionModifierPressed(event) ?
+    newSelection :
+    R.converge(
+      R.concat,
+      [
+        R.without(R.__, oldSelection),
+        R.without(oldSelection),
+      ]
+    )(newSelection)
+  );
+};
+
+// =============================================================================
+//
+// Marquee selection mode
+//
+// =============================================================================
+
+let patchSvgRef = null;
+
+const marqueeSelectingMode = {
+  getInitialState(props, { dragStartPosition, mousePosition }) {
+    return {
+      isMouseDown: true,
+      dragStartPosition,
+      mousePosition,
+      selection: props.selection,
+    };
+  },
+  onMouseMove(api, event) {
+    api.setState({
+      mousePosition: getMousePosition(patchSvgRef, api.props.offset, event),
+      selection: getComputedSelection(api, event),
+    });
+  },
+  onMouseUp(api, event) {
+    api.props.actions.setSelection(
+      getComputedSelection(api, event)
+    );
+    api.goToDefaultMode();
+  },
+  onKeyDownOrKeyUp(api, event) {
+    const oldSelection = api.state.selection;
+    const newSelection = getComputedSelection(api, event);
+
+    if (R.equals(oldSelection, newSelection)) return;
+
+    api.setState({
+      selection: newSelection,
+    });
+  },
+  getHotkeyHandlers(api) {
+    return {
+      [COMMAND.DESELECT]: api.goToDefaultMode,
+    };
+  },
+  renderMaquee(api) {
+    const { from, width, height } = getSelectionBox(api);
+    const inclusiveSelection = isInclusiveSelection(api);
+    const cls = className('MarqueeSelection', {
+      'is-inclusive': inclusiveSelection,
+    });
+
+    return (
+      <rect
+        className={cls}
+        x={from.x}
+        y={from.y}
+        width={width}
+        height={height}
+      />
+    );
+  },
+  renderBlockingLayer(api) {
```

Comment:
A comment about what exactly this layer blocks(patch entities from mouse events?) would be nice